README.md
# koa-http2-proxy

Configure [http2-proxy](https://github.com/nxtedition/node-http2-proxy) middleware with ease for [koa](https://github.com/koajs/koa).

Based on [http-proxy-middleware](https://github.com/chimurai/http-proxy-middleware).

## TL;DR

Proxy requests to `http://www.example.org`

```javascript
var Koa = require('koa');
var proxy = require('koa-http2-proxy');
var app = new Koa();

// response
app.use(proxy({ target: 'http://www.example.org' }));

app.listen(3000);

// http://localhost:3000/foo/bar -> http://www.example.org/foo/bar
```

:bulb: **Tip:** Set the option `changeOrigin` to `true` for [name-based virtual hosted sites](http://en.wikipedia.org/wiki/Virtual_hosting#Name-based).

## Table of Contents

<!-- MarkdownTOC autolink=true bracket=round depth=2 -->

- [Install](#install)
- [Core concept](#core-concept)
- [Example](#example)
- [Context matching](#context-matching)
- [Options](#options)
- [Shorthand](#shorthand)
  - [app.use\(path, proxy\)](#appusepath-proxy)
- [WebSocket](#websocket)
  - [External WebSocket upgrade](#external-websocket-upgrade)
- [Tests](#tests)
- [Changelog](#changelog)
- [License](#license)

<!-- /MarkdownTOC -->

## Install

```javascript
$ npm install --save-dev koa-http2-proxy
```

## Core concept

Proxy middleware configuration.

#### proxy([context,] config)

```javascript
var proxy = require('koa-http2-proxy');

var apiProxy = proxy('/api', { target: 'http://www.example.org' });
//                   \____/   \_____________________________/
//                     |                    |
//                   context             options

// 'apiProxy' is now ready to be used as middleware in a server.
```

- **context**: Determine which requests should be proxied to the target host.
  (more on [context matching](#context-matching))
- **options.target**: target host to proxy to. _(protocol + host)_

(full list of [`koa-http2-proxy` configuration options](#options))

#### proxy(uri [, config])

```javascript
// shorthand syntax for the example above:
var apiProxy = proxy('http://www.example.org/api');
```

More about the [shorthand configuration](#shorthand).

## Example

```javascript
// include dependencies
var Koa = require('koa');
var proxy = require('koa-http2-proxy');

// proxy middleware options
var options = {
  target: 'http://www.example.org', // target host
  ws: true, // proxy websockets
  pathRewrite: {
    '^/api/old-path': '/api/new-path', // rewrite path
    '^/api/remove/path': '/path' // remove base path
  },
  router: {
    // when request.headers.host == 'dev.localhost:3000',
    // override target 'http://www.example.org' to 'http://localhost:8000'
    'dev.localhost:3000': 'http://localhost:8000'
  }
};

// create the proxy (without context)
var exampleProxy = proxy(options);

// mount `exampleProxy` in web server
var app = new Koa();
app.use(exampleProxy);
app.listen(3000);
```

## Context matching

Providing an alternative way to decide which requests should be proxied; In case you are not able to use the server's [`path` parameter](http://expressjs.com/en/4x/api.html#app.use) to mount the proxy or when you need more flexibility.

[RFC 3986 `path`](https://tools.ietf.org/html/rfc3986#section-3.3) is used for context matching.

```
         foo://example.com:8042/over/there?name=ferret#nose
         \_/   \______________/\_________/ \_________/ \__/
          |           |            |            |        |
       scheme     authority       path        query   fragment
```

- **path matching**

  - `proxy({...})` - matches any path, all requests will be proxied.
  - `proxy('/', {...})` - matches any path, all requests will be proxied.
  - `proxy('/api', {...})` - matches paths starting with `/api`

- **multiple path matching**

  - `proxy(['/api', '/ajax', '/someotherpath'], {...})`

- **wildcard path matching**

  For fine-grained control you can use wildcard matching. Glob pattern matching is done by _micromatch_. Visit [micromatch](https://www.npmjs.com/package/micromatch) or [glob](https://www.npmjs.com/package/glob) for more globbing examples.

  - `proxy('**', {...})` matches any path, all requests will be proxied.
  - `proxy('**/*.html', {...})` matches any path which ends with `.html`
  - `proxy('/*.html', {...})` matches paths directly under path-absolute
  - `proxy('/api/**/*.html', {...})` matches requests ending with `.html` in the path of `/api`
  - `proxy(['/api/**', '/ajax/**'], {...})` combine multiple patterns
  - `proxy(['/api/**', '!**/bad.json'], {...})` exclusion

  **Note**: In multiple path matching, you cannot use string paths and wildcard paths together.

- **custom matching**

  For full control you can provide a custom function to determine which requests should be proxied or not.

  ```javascript
  /**
   * @return {Boolean}
   */
  var filter = function(pathname, req) {
    return pathname.match('^/api') && req.method === 'GET';
  };

  var apiProxy = proxy(filter, { target: 'http://www.example.org' });
  ```

## Options

- **option.pathRewrite**: object/function, rewrite target's url path. Object-keys will be used as _RegExp_ to match paths.

  ```javascript
  // rewrite path
  pathRewrite: {'^/old/api' : '/new/api'}

  // remove path
  pathRewrite: {'^/remove/api' : ''}

  // add base path
  pathRewrite: {'^/' : '/basepath/'}

  // custom rewriting
  pathRewrite: function (path, req) { return path.replace('/api', '/base/api') }
  ```

- **option.router**: object/function, re-target `option.target` for specific requests.

  ```javascript
  // Use `host` and/or `path` to match requests. First match will be used.
  // The order of the configuration matters.
  router: {
      'integration.localhost:3000' : 'http://localhost:8001',  // host only
      'staging.localhost:3000'     : 'http://localhost:8002',  // host only
      'localhost:3000/api'         : 'http://localhost:8003',  // host + path
      '/rest'                      : 'http://localhost:8004'   // path only
  }

  // Custom router function
  router: function(req) {
      return 'http://localhost:8004';
  }
  ```

- **option.logLevel**: string, ['debug', 'info', 'warn', 'error', 'silent']. Default: `'info'`

- **option.logProvider**: function, modify or replace log provider. Default: `console`.

  ```javascript
  // simple replace
  function logProvider(provider) {
    // replace the default console log provider.
    return require('winston');
  }
  ```

  ```javascript
  // verbose replacement
  function logProvider(provider) {
    var logger = new (require('winston')).Logger();

    var myCustomProvider = {
      log: logger.log,
      debug: logger.debug,
      info: logger.info,
      warn: logger.warn,
      error: logger.error
    };
    return myCustomProvider;
  }
  ```

- **option.onError**: function, subscribe to http-proxy's `error` event for custom error handling.

  ```javascript
  function onError(err, ctx) {
    ctx.response.status = 500;
    ctx.response.body =
      'Something went wrong. And we are reporting a custom error message.';
  }
  ```

- **option.onProxyRes**: function, subscribe to http-proxy's `proxyRes` event.

  ```javascript
  function onProxyRes(proxyRes, ctx) {
    proxyRes.headers['x-added'] = 'foobar'; // add new header to response
    delete proxyRes.headers['x-removed']; // remove header from response
  }
  ```

- **option.onProxyReq**: function, subscribe to http-proxy's `proxyReq` event.

  ```javascript
  function onProxyReq(proxyReq, ctx) {
    // add custom header to request
    proxyReq.setHeader('x-added', 'foobar');
    // or log the req
  }
  ```

- **option.onUpgrade**: function, called before upgrading a websocket connection.

  ```javascript
  onUpgrade: async ctx => {
    // add session middleware to the websocket connection
    // see option.app
    await session(ctx, () => {});
  };
  ```

- **option.app**: koa app, used to generate a koa ctx to be used in onUpgrade. If left blank, a object containing only `req` will be used as context
- **option.headers**: object, adds [request headers](https://en.wikipedia.org/wiki/List_of_HTTP_header_fields#Request_fields). (Example: `{host:'www.example.org'}`)
- **option.target**: url string to be parsed with the url module
- **option.ws**: true/false: if you want to proxy websockets
- **option.xfwd**: true/false, adds x-forward headers
- **option.changeOrigin**: true/false, Default: false - changes the origin of the host header to the target URL
- **option.protocol**: Agent protocol ('http' or 'https'). Defaults to 'http'
- **option.proxyTimeout**: timeout (in millis) when proxy receives no response from target
- **option.proxyName**: Proxy name used for Via header

## Shorthand

Use the shorthand syntax when verbose configuration is not needed. The `context` and `option.target` will be automatically configured when shorthand is used. Options can still be used if needed.

```javascript
proxy('http://www.example.org:8000/api');
// proxy('/api', {target: 'http://www.example.org:8000'});

proxy('http://www.example.org:8000/api/books/*/**.json');
// proxy('/api/books/*/**.json', {target: 'http://www.example.org:8000'});

proxy('http://www.example.org:8000/api');
// proxy('/api', {target: 'http://www.example.org:8000'});
```

## WebSocket

```javascript
// verbose api
proxy('/', { target: 'http://echo.websocket.org', ws: true });

// shorthand
proxy('http://echo.websocket.org', { ws: true });

// shorter shorthand
proxy('ws://echo.websocket.org');
```

### External WebSocket upgrade

In the previous WebSocket examples, http-proxy-middleware relies on a initial http request in order to listen to the http `upgrade` event. If you need to proxy WebSockets without the initial http request, you can subscribe to the server's http `upgrade` event manually.

```javascript
var wsProxy = proxy('ws://echo.websocket.org');

var app = new Koa();
app.use(wsProxy);

var server = app.listen(3000);
server.on('upgrade', wsProxy.upgrade); // <-- subscribe to http 'upgrade'
```

## Tests

Run the test suite:

```bash
# install dependencies
$ yarn

# linting
$ yarn lint
$ yarn lint:fix

# building (compile typescript to js)
$ yarn build

# unit tests
$ yarn test

# code coverage
$ yarn cover
```

## Changelog

- [View changelog](https://github.com/ontola/koa-http2-proxy/blob/master/CHANGELOG.md)

## License

The MIT License (MIT)

Copyright for portions of this project are held by Steven Chim, 2015-2019 as part of [http-proxy-middleware](https://github.com/chimurai/http-proxy-middleware). All other copyright for this project are held by Ontola BV, 2019.


CHANGELOG.md
# Changelog

## [v0.0.5]

- Fix https proxy requests.

## [v0.0.4](https://github.com/ontola/koa-http2-proxy/releases/tag/v0.0.4)

- handle errors caused by socket.destroy();

## [v0.0.3](https://github.com/ontola/koa-http2-proxy/releases/tag/v0.0.3)

- add protocol option
- correctly set x-forwarded-proto header

## [v0.0.2](https://github.com/ontola/koa-http2-proxy/releases/tag/v0.0.2)

- add onUpgrade callback

## [v0.0.1](https://github.com/ontola/koa-http2-proxy/releases/tag/v0.0.1)

- forked project from [http-proxy-middleware](https://github.com/chimurai/http-proxy-middleware)
- altered for usage with Koa and http2


test/types.spec.ts
import middleware from '../src';
import { Options } from '../src/types';

describe('http-proxy-middleware TypeScript Types', () => {
  let options: Options;

  beforeEach(() => {
    options = {
      target: 'http://example.org'
    };
  });

  it('should create proxy with just options', () => {
    const proxy = middleware(options);
    expect(proxy).toBeDefined();
  });

  describe('HPM Filters', () => {
    it('should create proxy with path filter', () => {
      const proxy = middleware('/path', options);
      expect(proxy).toBeDefined();
    });

    it('should create proxy with glob filter', () => {
      const proxy = middleware(['/path/**'], options);
      expect(proxy).toBeDefined();
    });

    it('should create proxy with custom filter', () => {
      const proxy = middleware((path, req) => true, options);
      expect(proxy).toBeDefined();
    });

    it('should create proxy with manual websocket upgrade function', () => {
      const proxy = middleware((path, req) => true, options);
      expect(proxy.upgrade).toBeDefined();
    });
  });

  describe('http-proxy options', () => {
    it('should extend from http-proxy options', () => {
      options = {
        target: 'http://example',
        ws: true
      };
      expect(options).toBeDefined();
    });
  });

  describe('http-proxy-middleware options', () => {
    describe('pathRewrite', () => {
      it('should have pathRewrite Type with table', () => {
        options = { pathRewrite: { '^/from': '/to' } };
        expect(options).toBeDefined();
      });

      it('should have pathRewrite Type with function', () => {
        options = { pathRewrite: (path, req) => '/path' };
        expect(options).toBeDefined();
      });
    });

    describe('router', () => {
      it('should have router Type with table', () => {
        options = { router: { '^/from': '/to' } };
        expect(options).toBeDefined();
      });

      it('should have router Type with function', () => {
        options = { router: path => '/path' };
        expect(options).toBeDefined();
      });
    });

    describe('logLevel', () => {
      it('should have logLevel Type', () => {
        options = { logLevel: 'info' };
        expect(options).toBeDefined();
      });
    });

    describe('logProvider', () => {
      it('should have logProvider Type', () => {
        options = {
          logProvider: currentProvider => {
            return {
              log: () => {
                return;
              },
              debug: () => {
                return;
              },
              info: () => {
                return;
              },
              warn: () => {
                return;
              },
              error: () => {
                return;
              }
            };
          }
        };
        expect(options).toBeDefined();
      });
    });

    describe('HPM http-proxy events', () => {
      it('should have onError type', () => {
        // tslint:disable no-empty
        options = { onError: (err, req, res) => {} };
        expect(options).toBeDefined();
      });

      it('should have onProxyReq type', () => {
        // tslint:disable no-empty
        options = { onProxyReq: (proxyReq, req, res) => {} };
        expect(options).toBeDefined();
      });

      it('should have onProxyRes type', () => {
        // tslint:disable no-empty
        options = { onProxyRes: (proxyRes, req, res) => {} };
        expect(options).toBeDefined();
      });

      it('should have onProxyReqWs type', () => {
        options = {
          // tslint:disable no-empty
          onProxyReqWs: (proxyReq, req, socket, opts, head) => {}
        };
        expect(options).toBeDefined();
      });

      it('should have onOpen type', () => {
        // tslint:disable no-empty
        options = { onOpen: proxySocket => {} };
        expect(options).toBeDefined();
      });

      it('should have onClose type', () => {
        // tslint:disable no-empty
        options = { onClose: (res, socket, head) => {} };
        expect(options).toBeDefined();
      });
    });
  });
});


src/types.ts
import express from 'express';
import http from 'http';
import httpProxy from 'http-proxy';
import net from 'net';

export interface IRequest extends express.Request {}
export interface IResponse extends express.Response {}

export interface IRequestHandler extends express.RequestHandler {
  upgrade?: (req: IRequest, socket: net.Socket, head: any) => void;
}

export type Filter = FilterPath | FilterWildcards | FilterFn;
export type FilterFn = (pathname: string, req: IRequest) => boolean;
export type FilterPath = string;
export type FilterWildcards = string[];

export interface Options extends httpProxy.ServerOptions {
  pathRewrite?:
    | { [regexp: string]: string }
    | ((path: string, req: IRequest) => string);
  router?: { [hostOrPath: string]: string } | ((req: IRequest) => string);
  logLevel?: 'debug' | 'info' | 'warn' | 'error' | 'silent';
  logProvider?(provider: LogProvider): LogProvider;

  onError?(err: Error, req: IRequest, res: IResponse): void;
  onProxyRes?(
    proxyRes: http.ServerResponse,
    req: IRequest,
    res: IResponse
  ): void;
  onProxyReq?(
    proxyReq: http.ClientRequest,
    req: IRequest,
    res: IResponse
  ): void;
  onProxyReqWs?(
    proxyReq: http.ClientRequest,
    req: IRequest,
    socket: net.Socket,
    options: httpProxy.ServerOptions,
    head: any
  ): void;
  onOpen?(proxySocket: net.Socket): void;
  onClose?(res: IResponse, socket: net.Socket, head: any): void;
}

interface LogProvider {
  log: Logger;
  debug?: Logger;
  info?: Logger;
  warn?: Logger;
  error?: Logger;
}

type Logger = (...args: any[]) => void;


src/router.ts
import _ from 'lodash';
import { getInstance } from './logger';
const logger = getInstance();

export function getTarget(req, config) {
  let newTarget;
  const router = config.router;

  if (_.isPlainObject(router)) {
    newTarget = getTargetFromProxyTable(req, router);
  } else if (_.isFunction(router)) {
    newTarget = router(req);
  }

  return newTarget;
}

function getTargetFromProxyTable(req, table) {
  let result;
  const host = req.headers.host;
  const path = req.url;

  const hostAndPath = host + path;

  _.forIn(table, (value, key) => {
    if (containsPath(key)) {
      if (hostAndPath.indexOf(key) > -1) {
        // match 'localhost:3000/api'
        result = table[key];
        logger.debug('[HPM] Router table match: "%s"', key);
        return false;
      }
    } else {
      if (key === host) {
        // match 'localhost:3000'
        result = table[key];
        logger.debug('[HPM] Router table match: "%s"', host);
        return false;
      }
    }
  });

  return result;
}

function containsPath(v) {
  return v.indexOf('/') > -1;
}


src/http-proxy-middleware.ts
import express from 'express';
import httpProxy from 'http-proxy';
import _ from 'lodash';
import { createConfig } from './config-factory';
import * as contextMatcher from './context-matcher';
import * as handlers from './handlers';
import { getArrow, getInstance } from './logger';
import * as PathRewriter from './path-rewriter';
import * as Router from './router';
import { Filter, IRequest, IRequestHandler, IResponse, Options } from './types';

export class HttpProxyMiddleware {
  private logger = getInstance();
  private config;
  private wsInternalSubscribed = false;
  private proxyOptions: Options;
  private proxy: httpProxy;
  private pathRewriter;

  constructor(context: Filter | Options, opts?: Options) {
    this.config = createConfig(context, opts);
    this.proxyOptions = this.config.options;

    // create proxy
    this.proxy = httpProxy.createProxyServer({});
    this.logger.info(
      `[HPM] Proxy created: ${this.config.context}  -> ${this.proxyOptions.target}`
    );

    this.pathRewriter = PathRewriter.createPathRewriter(
      this.proxyOptions.pathRewrite
    ); // returns undefined when "pathRewrite" is not provided

    // attach handler to http-proxy events
    handlers.init(this.proxy, this.proxyOptions);

    // log errors for debug purpose
    this.proxy.on('error', this.logError);

    // https://github.com/chimurai/http-proxy-middleware/issues/19
    // expose function to upgrade externally
    this.middleware.upgrade = (req, socket, head) => {
      if (!this.wsInternalSubscribed) {
        this.handleUpgrade(req, socket, head);
      }
    };
  }

  // https://github.com/Microsoft/TypeScript/wiki/'this'-in-TypeScript#red-flags-for-this
  public middleware: IRequestHandler = async (
    req: IRequest,
    res: IResponse,
    next: express.NextFunction
  ) => {
    if (this.shouldProxy(this.config.context, req)) {
      const activeProxyOptions = this.prepareProxyRequest(req);
      this.proxy.web(req, res, activeProxyOptions);
    } else {
      next();
    }

    if (this.proxyOptions.ws === true) {
      // use initial request to access the server object to subscribe to http upgrade event
      this.catchUpgradeRequest((req.connection as any).server);
    }
  };

  private catchUpgradeRequest = server => {
    if (!this.wsInternalSubscribed) {
      server.on('upgrade', this.handleUpgrade);
      // prevent duplicate upgrade handling;
      // in case external upgrade is also configured
      this.wsInternalSubscribed = true;
    }
  };

  private handleUpgrade = (req: IRequest, socket, head) => {
    if (this.shouldProxy(this.config.context, req)) {
      const activeProxyOptions = this.prepareProxyRequest(req);
      this.proxy.ws(req, socket, head, activeProxyOptions);
      this.logger.info('[HPM] Upgrading to WebSocket');
    }
  };

  /**
   * Determine whether request should be proxied.
   *
   * @private
   * @param  {String} context [description]
   * @param  {Object} req     [description]
   * @return {Boolean}
   */
  private shouldProxy = (context, req: IRequest) => {
    const path = req.originalUrl || req.url;
    return contextMatcher.match(context, path, req);
  };

  /**
   * Apply option.router and option.pathRewrite
   * Order matters:
   *    Router uses original path for routing;
   *    NOT the modified path, after it has been rewritten by pathRewrite
   * @param {Object} req
   * @return {Object} proxy options
   */
  private prepareProxyRequest = (req: IRequest) => {
    // https://github.com/chimurai/http-proxy-middleware/issues/17
    // https://github.com/chimurai/http-proxy-middleware/issues/94
    req.url = req.originalUrl || req.url;

    // store uri before it gets rewritten for logging
    const originalPath = req.url;
    const newProxyOptions = _.assign({}, this.proxyOptions);

    // Apply in order:
    // 1. option.router
    // 2. option.pathRewrite
    this.applyRouter(req, newProxyOptions);
    this.applyPathRewrite(req, this.pathRewriter);

    // debug logging for both http(s) and websockets
    if (this.proxyOptions.logLevel === 'debug') {
      const arrow = getArrow(
        originalPath,
        req.url,
        this.proxyOptions.target,
        newProxyOptions.target
      );
      this.logger.debug(
        '[HPM] %s %s %s %s',
        req.method,
        originalPath,
        arrow,
        newProxyOptions.target
      );
    }

    return newProxyOptions;
  };

  // Modify option.target when router present.
  private applyRouter = (req: IRequest, options) => {
    let newTarget;

    if (options.router) {
      newTarget = Router.getTarget(req, options);

      if (newTarget) {
        this.logger.debug(
          '[HPM] Router new target: %s -> "%s"',
          options.target,
          newTarget
        );
        options.target = newTarget;
      }
    }
  };

  // rewrite path
  private applyPathRewrite = (req: IRequest, pathRewriter) => {
    if (pathRewriter) {
      const path = pathRewriter(req.url, req);

      if (typeof path === 'string') {
        req.url = path;
      } else {
        this.logger.info(
          '[HPM] pathRewrite: No rewritten path found. (%s)',
          req.url
        );
      }
    }
  };

  private logError = (err, req: IRequest, res: IResponse) => {
    const hostname =
      (req.headers && req.headers.host) || (req.hostname || req.host); // (websocket) || (node0.10 || node 4/5)
    const target =
      (this.proxyOptions.target as any).host || this.proxyOptions.target;
    const errorMessage =
      '[HPM] Error occurred while trying to proxy request %s from %s to %s (%s) (%s)';
    const errReference =
      'https://nodejs.org/api/errors.html#errors_common_system_errors'; // link to Node Common Systems Errors page

    this.logger.error(
      errorMessage,
      req.url,
      hostname,
      target,
      err.code || err,
      errReference
    );
  };
}


src/koa-http2-proxy.ts
import { request as requestHttp } from 'http';
import { request as requestHttps } from 'https';

import * as finalhandler from 'finalhandler';
import * as proxy from 'http2-proxy';
import * as url from 'url';

import { createConfig } from './config-factory';
import * as contextMatcher from './context-matcher';
import { getArrow, getInstance } from './logger';
import * as PathRewriter from './path-rewriter';
import * as Router from './router';

export class KoaHttp2Proxy {
  private logger = getInstance();
  private config;
  private wsInternalSubscribed = false;
  private proxyOptions;
  private pathRewriter;

  constructor(context, opts) {
    this.config = createConfig(context, opts);
    this.proxyOptions = this.config.options;

    this.pathRewriter = PathRewriter.createPathRewriter(
      this.proxyOptions.pathRewrite
    ); // returns undefined when "pathRewrite" is not provided

    this.logger.info(
      `[HPM] Proxy created: ${this.config.context}  -> ${this.proxyOptions.target}`
    );
  }

  // https://github.com/Microsoft/TypeScript/wiki/'this'-in-TypeScript#red-flags-for-this
  public middleware = async (ctx, next) => {
    if (!this.shouldProxy(this.config.context, ctx.req)) {
      return next();
    }
    if (this.proxyOptions.ws === true) {
      // use initial request to access the server object to subscribe to http upgrade event
      this.catchUpgradeRequest(ctx.req);
    }

    return new Promise((resolve, reject) =>
      proxy.web(
        ctx.req,
        ctx.res,
        //@ts-ignore
        this.prepareProxyRequest(ctx, resolve),
        this.defaultWebHandler(ctx, resolve, reject)
      )
    );
  };

  private catchUpgradeRequest = req => {
    if (!this.wsInternalSubscribed) {
      req.connection.server.on('upgrade', this.handleUpgrade);
      // prevent duplicate upgrade handling;
      // in case external upgrade is also configured
      this.wsInternalSubscribed = true;
    }
  };

  private handleReq = ctx => (req, options) => {
    let proxyReq;
    const target = this.applyRouter(req, this.proxyOptions);
    const uri = url.parse(target);

    if (uri.protocol === 'http:') {
      proxyReq = requestHttp(options);
    } else {
      proxyReq = requestHttps(options);
    }
    

    if (!this.proxyOptions.changeOrigin) {
      proxyReq.setHeader('host', req.headers.host);
    }

    if (this.proxyOptions.headers) {
      for (const k of Object.keys(this.proxyOptions.headers)) {
        proxyReq.setHeader(k, this.proxyOptions.headers[k]);
      }
    }
    if (this.proxyOptions.xfwd) {
      proxyReq.setHeader('x-forwarded-for', req.socket.remoteAddress);
      const proto =
        req.headers['x-forwarded-proto'] ||
        (req.socket.encrypted ? 'https' : 'http');
      proxyReq.setHeader('x-forwarded-proto', proto);
      proxyReq.setHeader('x-forwarded-host', req.headers.host);
    }

    if (this.proxyOptions.onProxyReq) {
      this.proxyOptions.onProxyReq(proxyReq, ctx);
    }

    return proxyReq;
  };

  private handleRes = (ctx, resolve) => (_, __, proxyRes) => {
    if (this.proxyOptions.onProxyRes) {
      this.proxyOptions.onProxyRes(proxyRes, ctx);
    }

    ctx.response.status = proxyRes.statusCode;
    ctx.response.set(proxyRes.headers);
    ctx.response.body = proxyRes;

    if (resolve) {
      resolve();
    }
  };

  private handleUpgrade = async (req, socket, head) => {
    if (this.shouldProxy(this.config.context, req)) {
      const ctx = this.proxyOptions.app
        ? this.proxyOptions.app.createContext(req)
        : { req };
      if (this.proxyOptions.onUpgrade) {
        await this.proxyOptions.onUpgrade(ctx);
      }

      const activeProxyOptions = this.prepareProxyRequest(ctx, null);
      //@ts-ignore
      proxy.ws(req, socket, head, activeProxyOptions, this.defaultWSHandler);
      this.logger.info('[HPM] Upgrading to WebSocket');
    }
  };

  /**
   * Determine whether request should be proxied.
   *
   * @private
   * @param  {String} context [description]
   * @param  {Object} req     [description]
   * @return {Boolean}
   */
  private shouldProxy = (context, req) => {
    const path = req.originalUrl || req.url;
    return contextMatcher.match(context, path, req);
  };

  /**
   * Apply option.router and option.pathRewrite
   * Order matters:
   *    Router uses original path for routing;
   *    NOT the modified path, after it has been rewritten by pathRewrite
   * @param {Object} req
   * @return {Object} proxy options
   */
  private prepareProxyRequest = (ctx, resolve) => {
    // https://github.com/chimurai/http-proxy-middleware/issues/17
    // https://github.com/chimurai/http-proxy-middleware/issues/94
    ctx.req.url = ctx.req.originalUrl || ctx.req.url;

    // store uri before it gets rewritten for logging
    const originalPath = ctx.req.url;

    // Apply in order:
    // 1. option.router
    // 2. option.pathRewrite
    const target = this.applyRouter(ctx.req, this.proxyOptions);
    this.applyPathRewrite(ctx.req, this.pathRewriter);

    // debug logging for both http(s) and websockets
    if (this.proxyOptions.logLevel === 'debug') {
      const arrow = getArrow(
        originalPath,
        ctx.req.url,
        this.proxyOptions.target,
        target
      );
      this.logger.debug(
        '[HPM] %s %s %s %s',
        ctx.req.method,
        originalPath,
        arrow,
        target
      );
    }

    const uri = url.parse(target);

    return {
      hostname: uri.hostname,
      onReq: this.handleReq(ctx),
      onRes: this.handleRes(ctx, resolve),
      path: ctx.req.url,
      port: uri.port,
      protocol: uri.protocol,
      proxyName: this.proxyOptions.proxyName,
      proxyTimeout: this.proxyOptions.proxyTimeout,
      target
    };
  };

  // Modify option.target when router present.
  private applyRouter = (req, options) => {
    let newTarget;

    if (options.router) {
      newTarget = Router.getTarget(req, options);

      if (newTarget) {
        this.logger.debug(
          '[HPM] Router new target: %s -> "%s"',
          options.target,
          newTarget
        );
        return newTarget;
      }
    }

    return options.target;
  };

  // rewrite path
  private applyPathRewrite = (req, pathRewriter) => {
    if (pathRewriter) {
      const path = pathRewriter(req.url, req);

      if (typeof path === 'string') {
        req.url = path;
      } else {
        this.logger.info(
          '[HPM] pathRewrite: No rewritten path found. (%s)',
          req.url
        );
      }
    }
  };

  private defaultWebHandler = (ctx, resolve, reject) => (err, req, res) => {
    if (err) {
      this.logError(err, req);

      if (this.proxyOptions.onError) {
        this.proxyOptions.onError(err, ctx);
        resolve();
      } else {
        finalhandler(req, res)(err);
        reject();
      }
    } else {
      resolve();
    }
  };

  private defaultWSHandler = (err, req, socket) => {
    if (err) {
      this.logError(err, req);
      try {
        socket.destroy();
      } catch (e) {
        this.logError(e, req);
      }
    }
  };

  private logError = (err, req) => {
    const hostname =
      (req.headers && req.headers.host) || (req.hostname || req.host); // (websocket) || (node0.10 || node 4/5)
    const target = this.proxyOptions.target.host || this.proxyOptions.target;
    const errorMessage =
      '[HPM] Error occurred while trying to proxy request %s from %s to %s (%s) (%s)';
    const errReference =
      'https://nodejs.org/api/errors.html#errors_common_system_errors'; // link to Node Common Systems Errors page

    this.logger.error(
      errorMessage,
      req.url,
      hostname,
      target,
      err.code || err,
      errReference
    );
  };
}

function middleware(context, opts) {
  return new KoaHttp2Proxy(context, opts).middleware;
}

export default middleware;
